test/controllers/tasks_controller_test.rb

```diff
@@ -90,37 +85,100 @@
   # Unskip and complete these tests for Wave 3
   describe "edit" do
     it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+
+      # Act
+      get edit_task_path(task.id)
+
+      # Assert
+      must_respond_with :success
     end
 
-    it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+    it "will respond with redirect when attempting to edit a nonexistent task" do
+
+      # Act
+      get edit_task_path(-200)
+
+      # Assert
+      must_respond_with :redirect
     end
   end
 
   # Uncomment and complete these tests for Wave 3
   describe "update" do
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
+
     it "can update an existing task" do
-      # Your code here
+      get update_task_path(task.id)
+      must_respond_with :success
     end
 
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      get update_task_path(-200)
+
+      # Assert
+      must_respond_with :redirect
     end
   end
 
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it "redirects to tasks index page and deletes no tasks if no tasks exist" do
+      Task.destroy_all
+      invalid_task_id = 1
+
+      expect {
+        delete task_path(invalid_task_id)
+      }.must_differ "Task.count", 0
+
+      must_redirect_to tasks_path
+    end
+
+    it "redirects to tasks index page and deletes no tasks if deleting a task with an id that has already been deleted" do
+      Task.create(name: "Sample", description: "Sample Description")
+      task_id = Task.find_by(name: "Sample").id
+      Task.destroy_all
+
+      expect {
+        delete task_path(task_id)
+      }.must_differ "Task.count", 0
+
+      must_redirect_to tasks_path
+    end
 
+    it "decreases the task list count by one" do
+      test_task = Task.find_by(id: task[:id])
+
+      expect { delete task_path(Task.all.first.id) }.must_differ "Task.count", -1
+    end
+    it "will redirect if given an invalid id" do
+      get update_task_path(-200)
+
+      # Assert
+      must_respond_with :redirect
+    end
   end
 
   # Complete for Wave 4
   describe "toggle_complete" do
     # Your tests go here
+    it "returns a datetime when we mark a task completed and redirects to the root path" do
```

Comment:
What about the case of `toggle_complete` for a task that does not exist/invalid task id?